Source/WebKit/UIProcess/API/glib/WebKitImage.cpp

```diff
@@ -0,0 +1,477 @@
+{
+    auto* image = WEBKIT_IMAGE(object);
+
+    switch (propId) {
+    case PROP_WIDTH:
+        image->priv->width = g_value_get_int(value);
+        break;
+    case PROP_HEIGHT:
+        image->priv->height = g_value_get_int(value);
+        break;
+    case PROP_STRIDE:
+        image->priv->stride = g_value_get_uint(value);
+        break;
+    default:
+        G_OBJECT_WARN_INVALID_PROPERTY_ID(object, propId, paramSpec);
+    }
+}
+
+static void webkitImageGetProperty(GObject* object, guint propId, GValue* value, GParamSpec* paramSpec)
+{
+    auto* image = WEBKIT_IMAGE(object);
+
+    switch (propId) {
+    case PROP_WIDTH:
+        g_value_set_int(value, webkit_image_get_width(image));
+        break;
+    case PROP_HEIGHT:
+        g_value_set_int(value, webkit_image_get_height(image));
+        break;
+    case PROP_STRIDE:
+        g_value_set_uint(value, webkit_image_get_stride(image));
+        break;
+    default:
+        G_OBJECT_WARN_INVALID_PROPERTY_ID(object, propId, paramSpec);
+    }
+}
+
+static void webkitImageConstructed(GObject* object)
+{
+    G_OBJECT_CLASS(webkit_image_parent_class)->constructed(object);
+
+    auto* image = WEBKIT_IMAGE(object);
+    auto* priv = image->priv;
+
+    gsize rowBytes = priv->width * static_cast<uint8_t>(ImageFormatBPP::RGBA8);
+    g_return_if_fail(priv->stride >= rowBytes);
+}
+
+static void webkit_image_class_init(WebKitImageClass* imageClass)
+{
+    GObjectClass* objectClass = G_OBJECT_CLASS(imageClass);
+    objectClass->set_property = webkitImageSetProperty;
+    objectClass->get_property = webkitImageGetProperty;
+    objectClass->constructed = webkitImageConstructed;
+
+    /**
+     * WebKitImage:width:
+     *
+     * The image width in pixels
+     *
+     * Since: 2.52
+     */
+    sObjProperties[PROP_WIDTH] =
+    g_param_spec_int(
+        "width",
+        nullptr, nullptr,
+        1, G_MAXINT, 1,
+        static_cast<GParamFlags>(WEBKIT_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY));
+
+    /**
+     * WebKitImage:height:
+     *
+     * The image height
+     *
+     * Since: 2.52
+     */
+    sObjProperties[PROP_HEIGHT] =
+    g_param_spec_int(
+        "height",
+        nullptr, nullptr,
+        1, G_MAXINT, 1,
+        static_cast<GParamFlags>(WEBKIT_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY));
+
+    /**
+     * WebKitImage:stride:
+     *
+     * The image stride
+     *
+     * Since: 2.52
+     */
+    sObjProperties[PROP_STRIDE] =
+    g_param_spec_uint(
+        "stride",
+        nullptr, nullptr,
+        static_cast<uint8_t>(ImageFormatBPP::RGBA8), G_MAXUINT, 4,
+        static_cast<GParamFlags>(WEBKIT_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY));
+
+    g_object_class_install_properties(objectClass, N_PROPERTIES, sObjProperties.data());
+}
+
+WebKitImage* webkitImageNew(int width, int height, guint stride, CompletionHandler<GRefPtr<GBytes>(WebKitImage*)>&& createBytesFn)
+{
+    g_return_val_if_fail(createBytesFn, nullptr);
+
```

Comment:
We could actually move the assertion from the `.constructed` vfunc here, and remove the `webkitImageConstructed` function altogether. Also, as this is an internal WebKit function, let's replace `g_return_val_if_fail()` with an assertion:


```suggestion
    RELEASE_ASSERT(priv->width * RGBA8BytesPerPixel <= priv->stride);
    RELEASE_ASSERT(createBytesFn);
```

Reply:
my reasoning, when I added  the (incorrect) assertion to `.constructed` was that I wanted it to run for both `webkitImageNew` or `g_object_new` users. 

I'm happy to add the `RELEASE_ASSERT(createBytesFn)` here, I think I should add `RELEASE_ASSERT(priv->width * RGBA8BytesPerPixel <= priv->stride);` to the constructed, and keep it. Unless you are quite against the idea of the `constructed` just for that?

Reply:
I think it's better to remove the `webkitImageConstructed` function.

It would make sense if we would be expecting client code using the public API to create `WebKitImage` instances; but only WebKit will be doing that, and we can remind people to use the `webkitImageNew()` helper function during code review.